### ISSUE TITLE
Undeterministic behavior of ylcIf

### ISSUE THREAD

**Issue by ciolben** [pr83/ylc#5]
Hello!

I have this issue sometimes (if the ylcIf is working, and after reloading the page, it may not). The workaround is to generate an id based on the loop index and use jquery to show/hide when clicked.

version: 0.99.13, Chrome

Simplified Example: (I preserved the context)
```html
<table>
  <thead>
    <tr>
      <th>Name</th>
      <th>Content</th>
    </tr>
  </thead>
  <tbody>
    <tr data-ylcLoop="device: result">
      <td data-ylcBind="text <- device.name "></td>
      <td>
        <span data-ylcIf="device.expanded" data-ylcEvents="click: clickRow(device)">+</span>
        <table data-ylcIf="!device.expanded">
          <thead>
            <tr>
              <th><span data-ylcEvents="click: clickRow(device)">-</span>P1</th>
              <th>P2</th>
            </tr>
          </thead>
          <tbody>
            <tr data-ylcLoop="content: device.contents">
              <td data-ylcBind="text <- content.p1"></td>
              <td data-ylcBind="text <- content.p2"></td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  </tbody>
</table>
```

The clickRow(device) function is simply:
```javascript
if (typeof device.expanded === 'undefined') { device.expanded = false; }
 device.expanded = !device.expanded;
```

Any ideas why it is not consistent?
My idea: sometimes the events are evaluated before the ylcIf (in that case it works), but sometimes the other way around (and then we are in the wrong state between the model and the view).

Thank you!

**Comment by pr83** [comment]
Hello,

I have been trying to replicate it, but I suspect I need a bit more context.

I have tried it using this test code (tried to guess what your controller might look like):

<!DOCTYPE html>
<html>

	<head>
		<script src="http://code.jquery.com/jquery-1.11.1.min.js"></script>
		<script src="yellowCode.js"></script>
	</head>

	<body>

		<h1>Basic example</h1>
		<div id="hook">
			<table border="1">
			  <thead>
				<tr>
				  <th>Name</th>
				  <th>Content</th>
				</tr>
			  </thead>
			  <tbody>
				<tr data-ylcLoop="device: result">
				  <td data-ylcBind="text <- device.name "></td>
				  <td>
					<span data-ylcIf="!device.expanded" data-ylcEvents="click: clickRow(device)">+</span>
					<table data-ylcIf="device.expanded">
					  <thead>
						<tr>
						  <th><span data-ylcEvents="click: clickRow(device)">-</span>P1</th>
						  <th>P2</th>
						</tr>
					  </thead>
					  <tbody>
						<tr data-ylcLoop="content: device.contents">
						  <td data-ylcBind="text <- content.p1"></td>
						  <td data-ylcBind="text <- content.p2"></td>
						</tr>
					  </tbody>
					</table>
				  </td>
				</tr>
			  </tbody>
			</table>
		</div>

		<script>

			var controller = {
				init: function(model) {
					model.result = [
						{
							name: "Test 1",
							contents: [
								{
									p1: "test 1 a",
									p2: "test 1 b"
								}
							]
						},
						{
							name: "Test 2",
							contents: [
								{
									p1: "test 2 a",
									p2: "test 2 b"
								}
							]
						}
					];
				},
				
				clickRow: function(model, context, device) {
					if (typeof device.expanded === 'undefined') {
						device.expanded = false;
					}
					device.expanded = !device.expanded;
				}
			};

			$("#hook").yellowCode(controller);

		</script>

	</body>

</html>

I couldn't replicate it however.

Would you mind modifying the above code making the minimum necessary changes to make the issue manifest itself? Also, how often does it happen (how many times on average do you need to refresh the browser page)?

Thanks.

**Comment by ciolben** [comment]
You're right, the example I gave is not good because it's too simple and doesn't reflect the controller behavior (sorry for that).

Starting from the example, it always work when the page is loaded.
Then there is a button to get results from an ajax query which fills the array `model.result`.
The strange behavior begin here. However, it is not that strange anymore (continue reading).
I think I found why:
If I modify the view like this
```html
<tr data-ylcLoop="device, resultloop: result">
```
And the controller like this:
```javascript
device = model.result[context.loopStatuses.resultloop.index]; 
if (typeof device.expanded === 'undefined') { device.expanded = false; }
 device.expanded = !device.expanded;
```
Now it always works. That means the object `device` is not pointing to the current one passed in argument after updating the looped array. I think I need to precise that the `model.result` is becoming a new array (I don't reuse the old reference): `model.result = response.result;`.
Here I realize that I might be wrong with the term underterministic because I was sometimes using the same array reference, and sometimes not (my choices were underterministics :) ). However it is still not in sync with the model.

To reproduce:
You can add to the view:
```html
	  <button data-ylcEvents="click: clickGetResult">GetResult</button>
```
and to the controller:
```javascript
clickGetResult: function(model, context) {
                            //mimic an ajax call (to stay in the same context as mine)
			    context.updateModel(function(model){
  			    model.result = [
  					{
  						name: "Test 3",
  						contents: [
  							{
  								p1: "test 1 a",
  								p2: "test 1 b"
  							}
  						]
  					},
  					{
  						name: "Test 4",
  						contents: [
  							{
  								p1: "test 2 a",
  								p2: "test 2 b"
  							}
  						]
  					}
  				  ];
			    });
			  },
```

**Comment by pr83** [comment]
Thanks, now it makes perfect sense. I will have a look why it uses a stale underlying list.

**Comment by pr83** [comment]
Hello, thanks for the clarification in #5.

I have stripped down your test case to the bare minimum to show what the issue is:

```
<html>

	<head>
		<script src="http://code.jquery.com/jquery-1.11.1.min.js"></script>
		<script src="yellowCode.js"></script>
	</head>

	<body>
	
		<div id="hook">
			<button data-ylcEvents="click: updateValue">Set value to "New"</button>
			
			<div data-ylcLoop="element: result">
				with ylcIf:
				<a href="#" data-ylcIf="!(element.someFlag)"
				            data-ylcBind="text <- element.name"
							data-ylcEvents="click: alertName(element)"></a>
				<br/>
				without ylcIf:
				<a href="#" data-ylcBind="text <- element.name"
				            data-ylcEvents="click: alertName(element)"></a>
			</div>

		</div>

		<script>

			var controller = {
				init: function(model) {
					model.result = [
						{
							name: "Old",
							someFlag: false
						}
					];
				},
				
				alertName: function(model, context, element) {
					alert(element.name);
					return false;
				},
				
				updateValue: function(model, context) {
					model.result = [
						{
							name: "New",
							someFlag: false
						}
					];
				}
			}

			$("#hook").yellowCode(controller);

		</script>

	</body>

</html>
```

So basically the issue manifests itself when we refer to a list from a ylcEvents event handler parameter, the element with the ylcEvents also contains ylcIf and we replace the list completely.

**Comment by pr83** [comment]
Hello, I have committed a fix, please see release 0.99.21.

**Comment by ciolben** [comment]
The fix works perfectly for me.
Thank you very much for your work!
